File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        li1 = []
        current1 = l1
        while current1:
            li1.append(current1.val)
            current1 = current1.next
        li2 = []
        current2 = l2
        while current2:
            li2.append(current2.val)
            current2 = current2.next

        
        ln1 = 0
        for i in range(len(li1)):
            ln1+=li1[i]*10**i

        ln2 = 0
        for i in range(len(li2)):
            ln2+=li2[i]*10**i

        ln3=str(ln1+ln2)
        ln3 = ln3[::-1]
        ans = []
        for i in range(len(ln3)):
            ans.append(int(ln3[i]))

        
        def list_to_listnode(lst):
            if not lst:
                return None
            head = ListNode(lst[0])
            current = head
            for val in lst[1:]:
                current.next = ListNode(val)
                current = current.next
            return head

        return list_to_listnode(ans)

                
        

        

File: 0002-add-two-numbers/0002-add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
typedef struct ListNode ListNode;
ListNode * genListNode(int val, ListNode * next){
    ListNode *node = (ListNode *)malloc(sizeof(ListNode));
    node->val = val;
    node->next = next;
    return node;
}
ListNode * re_genListNode(int val, ListNode * current){
    ListNode * last = genListNode(val, NULL);
    ListNode*ptr = current;
    if(current == NULL) return last;
    while(ptr!=NULL){
        if(ptr->next==NULL){
            ptr->next = last;
            break;
        }
        ptr = ptr->next;
    }
    return current;
}
ListNode* reverseList(ListNode* head) {
    ListNode* prev = NULL;
    ListNode* curr = head;
    ListNode* next = NULL;

    while (curr != NULL) {
        // Store the next node
        next = curr->next;

        // Reverse the current node's pointer
        curr->next = prev;

        // Move pointers one position ahead
        prev = curr;
        curr = next;
    }

    // Update the head to the new first node
    head = prev;
    return head;
}

struct ListNode * addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    ListNode *ans;
    ListNode * previous = NULL;
    while(l1!=NULL && l2!=NULL){
        if(l1->val + l2->val <=9){
            ans = re_genListNode(l1->val + l2->val, previous);
            previous = ans;
        }else{
            ans = re_genListNode((l1->val + l2->val)%10, previous);
            previous = ans;
            if(l1->next!=NULL) l1->next->val+=1;
            if(l1->next==NULL){
                ListNode*temp = NULL;
                l1->next = genListNode(1, temp);
            }
        }
        

        if(l1->next != NULL && l2->next == NULL){
            ListNode * temp = NULL;
            l2->next = genListNode(0, temp);
        }
        if(l2->next != NULL && l1->next == NULL){
            ListNode * temp = NULL;
            l1->next = genListNode(0, temp);
        }
        l1 = l1->next;
        l2 = l2->next;
    }
    return ans;
}